src/pandapipes/pipeflow.py

```diff
@@ -296,67 +304,50 @@ def set_damping_factor(net, niter, error):
     current_alpha = get_net_option(net, "alpha")
-
-    if error_x0_increased and error_x1_increased:
+    if np.all(error_increased):
         set_net_option(net, "alpha", current_alpha / 10 if current_alpha >= 0.1 else current_alpha)
     else:
         set_net_option(net, "alpha", current_alpha * 10 if current_alpha <= 0.1 else 1.0)
-
-    return error_x0_increased, error_x1_increased
+    return error_increased
 
 
-def finalize_iteration(net, niter, error_1, error_2, residual_norm, nonlinear_method, tol_1, tol_2,
-                       tol_res, vals_1_old, vals_2_old, hydraulic_mode=True):
-    col1, col2 = (PINIT, MDOTINIT) if hydraulic_mode else (TINIT, TOUTINIT)
-
+def finalize_iteration(net, niter, residual_norm, nonlinear_method, errors, tols, tol_res, vals_old, vars, pit_names,
+                       diff):
     # Control of damping factor
     if nonlinear_method == "automatic":
-        error_x0_increased, error_x1_increased = set_damping_factor(net, niter,
-                                                                    [error_1, error_2])
-        if error_x0_increased:
-            net["_active_pit"]["node"][:, col1] = vals_1_old
-        if error_x1_increased:
-            net["_active_pit"]["branch"][:, col2] = vals_2_old
+        errors_increased = set_damping_factor(net, niter, errors)
+        logger.debug("alpha: %s" % get_net_option(net, "alpha"))
+        for error_increased, var, val, pit in zip(errors_increased, vars, vals_old, pit_names):
+            if error_increased:
+                net["_active_pit"][pit][:, globals()[var.upper() + 'INIT']] = val
+        if get_net_option(net, "alpha") != 1:
+            net.converged = False
+            return
+    elif nonlinear_method == "std":
+        max_std = np.std(list(diff.values()), axis=1)[:5].max()
+        alpha = 0.33 if max_std == 0 else 1 if max_std < 1 else 0.33 if max_std > 3 else 1 / max_std
+        set_net_option(net, "alpha", alpha)
     elif nonlinear_method != "constant":
         logger.warning("No proper nonlinear method chosen. Using constant settings.")
-
-    # Setting convergence flag
-    if error_2[niter] <= tol_2 and error_1[niter] <= tol_1 and residual_norm < tol_res:
-        if nonlinear_method != "automatic":
-            net.converged = True
-        elif get_net_option(net, "alpha") == 1:
-            net.converged = True
-
-    if hydraulic_mode:
-        logger.debug("errorv: %s" % error_1[niter])
-        logger.debug("errorp: %s" % error_2[niter])
-        logger.debug("alpha: %s" % get_net_option(net, "alpha"))
-    else:
-        logger.debug("errorT: %s" % error_1[niter])
-        logger.debug("alpha: %s" % get_net_option(net, "alpha"))
+    for error, var, tol in zip(errors.values(), vars, tols):
+        converged = error[niter] <= tol
+        if not converged: break
+        logger.debug("error_%s: %s" % (var, error[niter]))
+    net.converged = converged and residual_norm <= tol_res
```

Comment:
Should we discuss the value of tol_res? It is by default 1e-3 and it seems that it is hardly changed in any real world example. But does that value make sense? Is it realistic for all different types of examples and for each of the variables to be solved for? Mass flow can be extremely small in some cases, does 1e-3 make sense there?